### PR TITLE
[no-master] Fix #3408: Survive concurrent calls to `send` and `receive`.

### LINKED ISSUE #3408: Sporadic test failures in org.scalajs.testcommon.RPCCore (NullPointerException)
We're finding that our CI build fails about 10% of the time with a `ClosedException` caused by an NPE down in the guts. Has anyone run into this before? Any clues?

/cc @cquiroz 

```
Scala.js Com failed: Error: read ECONNRESET
  | [error] org.scalajs.testcommon.RPCCore$ClosedException
  | [error] 	at org.scalajs.testcommon.RPCCore.helpClose(RPCCore.scala:193)
  | [error] 	at org.scalajs.testcommon.RPCCore.call(RPCCore.scala:134)
  | [error] 	at org.scalajs.testcommon.RunMuxRPC.call(RunMuxRPC.scala:33)
  | [error] 	at org.scalajs.testadapter.RunnerAdapter.$anonfun$done$1(RunnerAdapter.scala:55)
  | [error] 	at scala.collection.immutable.List.map(List.scala:283)
  | [error] 	at org.scalajs.testadapter.RunnerAdapter.done(RunnerAdapter.scala:55)
  | [error] 	at sbt.Defaults$.$anonfun$allTestGroupsTask$6(Defaults.scala:936)
  | [error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
  | [error] 	at scala.collection.immutable.Map$Map2.foreach(Map.scala:146)
  | [error] 	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
  | [error] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
  | [error] 	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
  | [error] 	at sbt.Defaults$.$anonfun$allTestGroupsTask$5(Defaults.scala:934)
  | [error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
  | [error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:39)
  | [error] 	at sbt.std.Transform$$anon$4.work(System.scala:66)
  | [error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
  | [error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
  | [error] 	at sbt.Execute.work(Execute.scala:271)
  | [error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
  | [error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
  | [error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
  | [error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  | [error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  | [error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  | [error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  | [error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  | [error] 	at java.lang.Thread.run(Thread.java:748)
  | [error] Caused by: java.lang.NullPointerException
  | [error] 	at org.scalajs.jsenv.nodejs.AbstractNodeJSEnv$NodeComJSRunner.receive(AbstractNodeJSEnv.scala:257)
  | [error] 	at org.scalajs.jsenv.nodejs.AbstractNodeJSEnv$NodeComJSRunner.receive$(AbstractNodeJSEnv.scala:253)
  | [error] 	at org.scalajs.jsenv.nodejs.NodeJSEnv$ComNodeRunner.receive(NodeJSEnv.scala:70)
  | [error] 	at org.scalajs.jsenv.ComJSRunner.receive(ComJSRunner.scala:15)
  | [error] 	at org.scalajs.jsenv.ComJSRunner.receive$(ComJSRunner.scala:15)
  | [error] 	at org.scalajs.jsenv.nodejs.NodeJSEnv$ComNodeRunner.receive(NodeJSEnv.scala:70)
  | [error] 	at org.scalajs.testadapter.ComJSEnvRPC$$anon$1.run(ComJSEnvRPC.scala:31)
  | [error] java.lang.NullPointerException
  | [error] 	at org.scalajs.jsenv.nodejs.AbstractNodeJSEnv$NodeComJSRunner.send(AbstractNodeJSEnv.scala:247)
  | [error] 	at org.scalajs.jsenv.nodejs.AbstractNodeJSEnv$NodeComJSRunner.send$(AbstractNodeJSEnv.scala:245)
  | [error] 	at org.scalajs.jsenv.nodejs.NodeJSEnv$ComNodeRunner.send(NodeJSEnv.scala:70)
  | [error] 	at org.scalajs.testadapter.ComJSEnvRPC.send(ComJSEnvRPC.scala:67)
  | [error] 	at org.scalajs.testcommon.RPCCore.call(RPCCore.scala:137)
  | [error] 	at org.scalajs.testadapter.TestAdapter.loadFrameworks(TestAdapter.scala:53)
  | [error] 	at org.scalajs.sbtplugin.ScalaJSPluginInternal$.$anonfun$scalaJSTestFrameworkSettings$1(ScalaJSPluginInternal.scala:1012)
  | [error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
  | [error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:39)
  | [error] 	at sbt.std.Transform$$anon$4.work(System.scala:66)
  | [error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
  | [error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
  | [error] 	at sbt.Execute.work(Execute.scala:271)
  | [error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
  | [error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
  | [error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
  | [error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  | [error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  | [error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
  | [error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  | [error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  | [error] 	at java.lang.Thread.run(Thread.java:748)
 ```

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

File: js-envs-test-suite/src/test/scala/org/scalajs/jsenv/test/NodeJSTest.scala [hunk 1]
```diff
@@ -76,4 +76,35 @@ class NodeJSTest extends TimeoutComTests {
     com.await(DefaultTimeout)
   }
 
+  @Test
+  def testConcurrentSendReceive_issue3408: Unit = {
+    for (i <- 0 until 50) {
```
Comment (sjrd) [inline js-envs-test-suite/src/test/scala/org/scalajs/jsenv/test/NodeJSTest.scala]: 50 was enough to reliably reproduce the `NullPointerException` on each run.
I also stress-tested the fix with 500.

**Comment by sjrd** [comment]
I've also noticed that this PR, counter-intuitively, *improves* the startup time of `test`. And I think this can be correlated with https://github.com/scala-js/scala-js/pull/3354. I think what happened is that a `send()` and a `receive()` would concurrently enter the `while` loop, and *both* perform an `accept()`. One of them would immediately/quickly return with the connection, while the other one would hopelessly wait for a second connection to come in, until the timeout expired and it got the chance to realize that `comSocket` had been set by another thread.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala [hunk 2]
```diff
@@ -280,30 +289,52 @@ abstract class AbstractNodeJSEnv(
...
+          val js2jvm0 = new DataInputStream(
+              new BufferedInputStream(comSocket0.getInputStream()))
+
+          /* Assign those three fields together, without the possibility of an
+           * exception happening in the middle (see #3408).
+           */
+          comSocket = comSocket0
+          jvm2js = jvm2js0
+          js2jvm = js2jvm0
         }
```
Comment (gzm0) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala]: Shouldn't this go outside of the while loop? Have you considered putting the accept call in a tail recursive helper instead?

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala [hunk 3]
```diff
@@ -280,30 +289,52 @@ abstract class AbstractNodeJSEnv(
...
+      /* Now wait for a possibly still-successful `awaitConnection()` to
+       * complete before closing the sockets.
+       */
+      synchronized {
+        if (jvm2js != null)
+          jvm2js.close()
+        if (js2jvm != null)
+          js2jvm.close()
+        if (comSocket != null)
+          comSocket.close()
```
Comment (gzm0) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala]: All the checks don't seem necessary anymore; but maybe better leave them?

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala (hunk 2 above)
Comment (gzm0) [reply js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala]: I more meant something like this:

```scala
@tailrec
def waitAccept(): Option[Socket] = {
  if (!isRunning) {
    None
  } else {
    try {
      Some(serverSocket.accept())
    } catch {
      case _: SocketTimeoutException => waitAccept()
    }
  }
}

val optCom = waitAccept()

optCom.foreach {
  // Create streams and assign.
}

optCom.isDefined
```

IMO the exact conditions when what happens are clearer that way. Also the assignment is not in the loop anymore; it only can happen in the last iteration (if at all).

Comment (sjrd) [reply js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala]: OK, I see. Better that way, indeed. Done.

**Pull request merged**